File: js/src/utils/reverseLookup.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.reverseLookup = void 0;
const state_1 = require("../state");
const error_1 = require("../error");
const deserializeReverse_1 = require("./deserializeReverse");
const getReverseKeyFromDomainKey_1 = require("./getReverseKeyFromDomainKey");
/**
 * This function can be used to perform a reverse look up
 * @param connection The Solana RPC connection
 * @param nameAccount The public key of the domain to look up
 * @returns The human readable domain name
 */
function reverseLookup(connection, nameAccount, parent) {
    return __awaiter(this, void 0, void 0, function* () {
        const reverseKey = (0, getReverseKeyFromDomainKey_1.getReverseKeyFromDomainKey)(nameAccount, parent);
        const { registry } = yield state_1.NameRegistryState.retrieve(connection, reverseKey);
        if (!registry.data) {
            throw new error_1.NoAccountDataError("The registry data is empty");
        }
        return (0, deserializeReverse_1.deserializeReverse)(registry.data, !!parent);
    });
}
exports.reverseLookup = reverseLookup;

File: python/src/utils/reverse_lookup.py
from typing import Optional

from solana.rpc.async_api import AsyncClient
from solders.pubkey import Pubkey

from NameRegistryState import NameRegistryState
from exception import NoAccountDataException
from utils.deserialize_reverse import deserialize_reverse
from utils.get_reverse_key_from_domain_key import get_reverse_key_from_domain_key


async def reverse_lookup(
    connection: AsyncClient, name_account: Pubkey, parent: Optional[Pubkey] = None
) -> str:
    """
    Reverse lookup the name of an account given the address of the account.
    :param connection:
    :param name_account:
    :param parent:
    :return: human-readable name of the domain name
    """
    reverse_key = get_reverse_key_from_domain_key(name_account, parent)

    registry = await NameRegistryState.retrieve(connection, reverse_key)
    if registry is None or registry.data is None:
        raise NoAccountDataException("The registry data is empty")

    return deserialize_reverse(registry.data, bool(parent is not None))  # type: ignore[no-any-return]
